File: game-ui-harnesses/ui-component-harness/src/select-option-icons.ts
import type { Issue } from './contract.ts';
import type { Layout } from './tree-contract.ts';

/** Coordinates are relative to one equal-height row of popupContentLayout. */
export interface SelectOptionIcons<Reference extends 'image' | 'layerId' = 'image'> {
  version: '1.0';
  coordinateSpace: 'popup-row-local';
  items: Array<{
    optionId: string;
    icon: ({ layout: Layout } & Record<Reference, string>) | null;
    labelLayout: Layout;
  }>;
}

/** Shared producer-binding/runtime geometry validation. No IO or inferred links. */
export function validateSelectOptionIcons(
  value: unknown, optionIds: readonly string[], content: unknown,
  reference: 'image' | 'layerId', path: string,
  checkReference: (value: unknown, path: string) => void,
): Issue[] {
  const issues: Issue[] = [];
  const fail = (p: string, code: string, message: string) => issues.push({ path: p, code, message });
  const object = (v: unknown, p: string, keys: string[]): Record<string, unknown> | undefined => {
    if (!v || typeof v !== 'object' || Array.isArray(v)) { fail(p, 'OBJECT_REQUIRED', 'must be an object'); return; }
    const o = v as Record<string, unknown>;
    for (const k of Object.keys(o)) if (!keys.includes(k)) fail(`${p}.${k}`, 'UNKNOWN_FIELD', 'unsupported field');
    for (const k of keys) if (!Object.hasOwn(o, k)) fail(`${p}.${k}`, 'REQUIRED_FIELD', 'field is required');
    return o;
  };
  const data = object(value, path, ['version', 'coordinateSpace', 'items']);
  if (!data) return issues;
  if (data.version !== '1.0') fail(`${path}.version`, 'UNSUPPORTED_VERSION', 'only 1.0 is supported');
  if (data.coordinateSpace !== 'popup-row-local') fail(`${path}.coordinateSpace`, 'UNSUPPORTED_COORDINATE_SPACE', 'must be popup-row-local');
  const safe = content as Layout | undefined;
  if (!safe || ![safe.width, safe.height].every(n => Number.isFinite(n) && n > 0)) fail(path, 'POPUP_CONTENT_REQUIRED', 'optionIcons requires explicit valid popupContentLayout');
  const rowWidth = safe?.width ?? 0, rowHeight = (safe?.height ?? 0) / optionIds.length;
  const rect = (v: unknown, p: string): Layout | undefined => {
    const r = object(v, p, ['x', 'y', 'width', 'height']);
    if (!r) return;
    if (![r.x, r.y, r.width, r.height].every(n => typeof n === 'number' && Number.isFinite(n))
      || (r.x as number) < 0 || (r.y as number) < 0 || (r.width as number) <= 0 || (r.height as number) <= 0) {
      fail(p, 'INVALID_LAYOUT', 'finite nonnegative coordinates and positive dimensions required'); return;
    }
    const result = r as unknown as Layout;
    if (result.x + result.width > rowWidth + 1e-6 || result.y + result.height > rowHeight + 1e-6) fail(p, 'OPTION_LAYOUT_OUT_OF_BOUNDS', 'must fit inside its popup content row');
    return result;
  };
  if (!Array.isArray(data.items)) { fail(`${path}.items`, 'ARRAY_REQUIRED', 'must be an array'); return issues; }
  if (data.items.length !== optionIds.length) fail(`${path}.items`, 'OPTION_COVERAGE_REQUIRED', 'declare every option exactly once; icon:null explicitly means no icon');
  const seen = new Set<string>();
  data.items.forEach((raw, i) => {
    const p = `${path}.items[${i}]`, item = object(raw, p, ['optionId', 'icon', 'labelLayout']);
    if (!item) return;
    if (typeof item.optionId !== 'string' || !optionIds.includes(item.optionId)) fail(`${p}.optionId`, 'UNKNOWN_OPTION', 'must reference an option of this Select');
    else if (seen.has(item.optionId)) fail(`${p}.optionId`, 'DUPLICATE_OPTION', 'each option must appear once');
    else seen.add(item.optionId);
    const label = rect(item.labelLayout, `${p}.labelLayout`);
    if (item.icon === null) return;
    const icon = object(item.icon, `${p}.icon`, [reference, 'layout']);
    if (!icon) return;
    checkReference(icon[reference], `${p}.icon.${reference}`);
    const box = rect(icon.layout, `${p}.icon.layout`);
    if (label && box && label.x < box.x + box.width && label.x + label.width > box.x && label.y < box.y + box.height && label.y + label.height > box.y) fail(p, 'OPTION_ICON_TEXT_OVERLAP', 'icon and label reservations must not overlap');
  });
  return issues;
}

File: game-ui-harnesses/ui-decomposition-harness/src/ai_ui_decomposition/select_option_icons.py
"""Producer checks for the consumer's select-option-icons-v1 contract."""
import math

from .common import require


def _object(value, keys, code):
    require(isinstance(value, dict) and set(value) == set(keys), code)


def _rect(value, width, height):
    code = 'SELECT_OPTION_ICONS_LAYOUT_INVALID'
    _object(value, ('x', 'y', 'width', 'height'), code)
    require(all(type(v) in (int, float) and math.isfinite(v) for v in value.values()), code)
    require(value['x'] >= 0 and value['y'] >= 0 and value['width'] > 0 and value['height'] > 0, code)
    require(value['x'] + value['width'] <= width + 1e-6 and
            value['y'] + value['height'] <= height + 1e-6, code)


def validate_option_icons(state, option_ids, layers=None, popup_size=None, *, reference='layerId'):
    """Return explicit items, or None for a legacy declaration. No inferred icons."""
    if 'optionIcons' not in state:
        return None
    value = state['optionIcons']
    _object(value, ('version', 'coordinateSpace', 'items'), 'SELECT_OPTION_ICONS_INVALID')
    require(value['version'] == '1.0', 'SELECT_OPTION_ICONS_VERSION_UNSUPPORTED')
    require(value['coordinateSpace'] == 'popup-row-local', 'SELECT_OPTION_ICONS_COORDINATE_SPACE')
    safe = state.get('popupContentLayout')
    expected = ('x', 'y', 'width', 'height') if reference == 'image' else ('coordinateSpace', 'x', 'y', 'width', 'height')
    _object(safe, expected, 'SELECT_OPTION_ICONS_SAFE_AREA_REQUIRED')
    if reference == 'layerId':
        require(safe['coordinateSpace'] == 'target-popup-local', 'SELECT_OPTION_ICONS_COORDINATE_SPACE')
    rect = {k: safe[k] for k in ('x', 'y', 'width', 'height')}
    _rect(rect, *(popup_size or (float('inf'), float('inf'))))
    require(isinstance(option_ids, list) and len(option_ids) > 0 and all(isinstance(i, str) for i in option_ids) and
            len(set(option_ids)) == len(option_ids), 'SELECT_OPTION_ICONS_OPTIONS_INVALID')
    items = value['items']
    require(isinstance(items, list) and len(items) == len(option_ids), 'SELECT_OPTION_ICONS_OPTION_COVERAGE')
    seen = set()
    for item in items:
        _object(item, ('optionId', 'icon', 'labelLayout'), 'SELECT_OPTION_ICONS_ITEM_INVALID')
        ident = item['optionId']
        require(isinstance(ident, str) and ident in option_ids and ident not in seen, 'SELECT_OPTION_ICONS_OPTION_COVERAGE')
        seen.add(ident)
        label = item['labelLayout']
        _rect(label, safe['width'], safe['height'] / len(option_ids))
        icon = item['icon']
        if icon is None:
            continue
        _object(icon, (reference, 'layout'), 'SELECT_OPTION_ICONS_ICON_INVALID')
        require(isinstance(icon[reference], str) and bool(icon[reference].strip()), 'SELECT_OPTION_ICONS_REFERENCE_INVALID')
        require(layers is None or icon[reference] in layers, 'SELECT_OPTION_ICONS_UNKNOWN_LAYER')
        r = icon['layout']
        _rect(r, safe['width'], safe['height'] / len(option_ids))
        overlap = (r['x'] < label['x'] + label['width'] - 1e-6 and label['x'] < r['x'] + r['width'] - 1e-6 and
                   r['y'] < label['y'] + label['height'] - 1e-6 and label['y'] < r['y'] + r['height'] - 1e-6)
        require(not overlap, 'SELECT_OPTION_ICONS_LABEL_OVERLAP')
    return items


def contain_rect(layout, size, origin):
    """Use the whole PNG canvas including alpha padding, never an alpha bbox."""
    scale = min(layout['width'] / size[0], layout['height'] / size[1])
    width, height = size[0] * scale, size[1] * scale
    return [origin[0] + layout['x'] + (layout['width'] - width) / 2,
            origin[1] + layout['y'] + (layout['height'] - height) / 2, width, height]
